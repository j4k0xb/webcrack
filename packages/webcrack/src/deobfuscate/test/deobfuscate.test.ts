import { parse } from '@babel/parser';
import traverse from '@babel/traverse';
import { describe, expect, test } from 'vitest';
import {
  generate,
  inlineFunctionAliases,
  inlineVariableAliases,
} from '../../ast-utils';

describe('inline decoder', () => {
  test('inline variable', () => {
    const ast = parse(`
      function decoder() {}
      decoder(1);
      (() => {
        const alias = decoder, alias3 = alias;
        alias(2);
        alias3(3);
        (() => {
          let alias2;
          (alias2 = alias)(4);
        });
        let alias4;
        alias4 = alias;
        alias4(5);
      });
  `);
    traverse(ast, {
      FunctionDeclaration(path) {
        const binding = path.scope.getBinding('decoder')!;
        inlineVariableAliases(binding);
        path.stop();
      },
    });
    expect(generate(ast)).toMatchSnapshot();
  });

  test('inline function', () => {
    const ast = parse(`
      function decoder(a, b) {}
      decoder(1, 2);
      function ignore() {
        return decoder(3, 4);
      }
      (() => {
        function alias(a, b) {
          return decoder(a - 625, b);
        }
        alias(2, 3);
        (() => {
          function alias2(a, b) {
            return alias(b - -678, a);
          }
          alias2(4, 5);
        })();
      })();
  `);
    traverse(ast, {
      FunctionDeclaration(path) {
        const binding = path.scope.parent.bindings.decoder;
        inlineFunctionAliases(binding);
        path.stop();
      },
    });

    expect(generate(ast)).toMatchSnapshot();
  });
});
