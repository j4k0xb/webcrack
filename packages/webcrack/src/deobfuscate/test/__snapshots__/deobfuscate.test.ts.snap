// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`inline decoder > inline function 1`] = `
"function decoder(a, b) {}
decoder(1, 2);
function ignore() {
  return decoder(3, 4);
}
(() => {
  decoder(2 - 625, 3);
  (() => {
    decoder(5 - -678 - 625, 4);
  })();
})();"
`;

exports[`inline decoder > inline variable 1`] = `
"function decoder() {}
decoder(1);
() => {
  decoder(2);
  decoder(3);
  () => {
    decoder(4);
  };
  decoder(5);
};"
`;
